File: app/src/main/java/com/example/todolist/DbHandler.java
package com.example.todolist;

import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;
import java.util.List;

//Create a DB to store ToDo data
//https://developer.android.com/training/data-storage/sqlite
public class DbHandler extends SQLiteOpenHelper {
    private static final int VERSION = 1;
    private static final String NAME = "ToDoList.db";//Name of the database
    private static final String TABLE = "todo";//Name of the table
    private static final String ID = "id";//Primary key
    private static final String TASK = "task";//Text
    private static final String STATUS = "status";//Status
    private static final String CREATE_TABLE ="CREATE TABLE " +
    TABLE + "(" + ID + " INTEGER PRIMARY KEY AUTOINCREMENT, "
            + TASK + " TEXT, " +
                STATUS + " INTEGER)";
    private SQLiteDatabase db;

    public DbHandler(Context context) {
        super(context, NAME, null, VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db){
        db.execSQL(CREATE_TABLE);//executes SQL
    }
    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion){
        db.execSQL("DROP TABLE IF EXISTS " + TABLE);//Drops the older table
        onCreate(db);//adds the table back again with updated data
    }
    public void openDatabase(){
        db = this.getWritableDatabase();
    }
    public void addTask(ToDoItemModel task){
        ContentValues values = new ContentValues();
        values.put(TASK, task.getTask());
        values.put(STATUS, task.getStatus());
        db.insert(TABLE, null, values);
    }

    @SuppressLint("Range")
    public List<ToDoItemModel> getAllTasks(){
        List<ToDoItemModel> tasks = new ArrayList<>();
        Cursor cursor = null;
        db.beginTransaction();
        try{//Should return all rows from db
            cursor = db.query(TABLE,
                    null, null, null,
                    null, null, null, null);
            if(cursor != null){
                if(cursor.moveToFirst()){
                    do {
                        ToDoItemModel task = new ToDoItemModel();
                        task.setID(cursor.getInt(cursor.getColumnIndex(ID)));
                        task.setTask(cursor.getString(cursor.getColumnIndex(TASK)));
                        task.setStatus(cursor.getInt(cursor.getColumnIndex(STATUS)));
                        tasks.add(task);
                    } while(cursor.moveToNext());
                }
            }
        }
        finally {
            db.endTransaction();
            cursor.close();
        }
        return tasks;
    }

    public void updateTaskStatus(int id, int status){
        ContentValues values = new ContentValues();
        values.put(STATUS, status);
        db.update(TABLE, values, ID + "=?", new String[]{String.valueOf(id)});
    }
    public void updateTask(int id, String task){
        ContentValues values = new ContentValues();
        values.put(TASK, task);
        db.update(TABLE, values, ID + "=?", new String[]{String.valueOf(id)});
    }
    public void deleteTask(int id){
        db.delete(TABLE, ID + "=?", new String[]{String.valueOf(id)});
    }

}

File: app/src/main/java/com/example/todolist/DbHandler.kt
package com.example.todolist

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

class DbHandler(context: Context) : SQLiteOpenHelper(context, NAME, null, VERSION) {

    companion object {
        private const val VERSION = 1
        private const val NAME = "ToDoList.db"
        private const val TABLE = "todo"
        private const val ID = "id"
        private const val TASK = "task"
        private const val STATUS = "status"
        private const val CREATE_TABLE =
            "CREATE TABLE $TABLE ($ID INTEGER PRIMARY KEY AUTOINCREMENT, $TASK TEXT, $STATUS INTEGER)"
    }

    private var db: SQLiteDatabase? = null

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(CREATE_TABLE)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS $TABLE")
        onCreate(db)
    }

    fun openDatabase() {
        db = this.writableDatabase
    }

    fun addTask(task: ToDoItemModel) {
        val values = ContentValues().apply {
            put(TASK, task.task)
            put(STATUS, task.status)
        }
        db?.insert(TABLE, null, values)
    }

    @SuppressLint("Range")
    fun getAllTasks(): MutableList<ToDoItemModel> {
        val tasks = mutableListOf<ToDoItemModel>()
        val cursor = db?.query(TABLE, null, null, null,
            null, null, null)

        cursor?.use {
            if (it.moveToFirst()) {
                do {
                    val task = ToDoItemModel().apply {
                        id = it.getInt(it.getColumnIndex(ID))
                        task = it.getString(it.getColumnIndex(TASK))
                        status = it.getInt(it.getColumnIndex(STATUS))
                    }
                    tasks.add(task)
                } while (it.moveToNext())
            }
        }

        return tasks
    }

    fun updateTaskStatus(id: Int, status: Int) {
        val values = ContentValues().apply {
            put(STATUS, status)
        }
        db?.update(TABLE, values, "$ID=?", arrayOf(id.toString()))
    }

    fun updateTask(id: Int, task: String) {
        val values = ContentValues().apply {
            put(TASK, task)
        }
        db?.update(TABLE, values, "$ID=?", arrayOf(id.toString()))
    }

    fun deleteTask(id: Int) {
        db?.delete(TABLE, "$ID=?", arrayOf(id.toString()))
    }
}